core/src/main/java/com/linecorp/armeria/server/RedirectService.java

```diff
@@ -128,16 +247,36 @@ public RedirectService(HttpStatus redirectStatus,
         }
         httpStatus = redirectStatus;
         this.locationFunction = locationFunction;
+        this.preserveQueryString = preserveQueryString;
     }
 
     /**
      * NB: For now we redirect all methods.
      */
     @Override
-    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req)
-            throws Exception {
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        String location = locationFunction.apply(ctx);
+        requireNonNull(location, "locationFunction returned null.");
+
+        if (preserveQueryString) {
+            location = appendQueryString(ctx, location);
+        }
+
         return HttpResponse.of(HttpHeaders.of(httpStatus)
-                .set(HttpHeaderNames.LOCATION, locationFunction.apply(ctx)));
+                                          .set(HttpHeaderNames.LOCATION, location));
+    }
+
+    private static String appendQueryString(ServiceRequestContext ctx, String location) {
+        final String query = ctx.query();
+        if (query == null || location.lastIndexOf('?') >= 0) {
+            // The request URI does not have a query string or
+            // The new location includes a query string already.
+            return location;
+        }
+
+        final StringBuilder buf = new StringBuilder(location.length() + query.length() + 1);
```

Comment:
Since there isn't any conditional logic, concatenation should be at least as fast as this.

Reply:
Oh, does simple concatenation perform as well as or better than this?

Reply:
Yeah in Java 8 it should be the same bytecode AFAIU. In future Java, since they can customize concatenation bytecode using invokedynamic it may get even better